gems/aws-sdk-s3/lib/aws-sdk-s3/directory_downloader.rb

```diff
@@ -0,0 +1,155 @@
+# frozen_string_literal: true
+
+module Aws
+  module S3
+    # Raised when DirectoryDownloader fails to download objects from S3 bucket
+    class DirectoryDownloadError < StandardError
+      def initialize(message, errors = [])
+        @errors = errors
+        super(message)
+      end
+
+      # @return [Array<StandardError>] The list of errors encountered when downloading objects
+      attr_reader :errors
+    end
+
+    # @api private
+    class DirectoryDownloader
+      def initialize(options = {})
+        @client = options[:client] || Client.new
+        @executor = options[:executor] || DefaultExecutor.new
+        @options = options
+        @abort_download = false
+      end
+
+      attr_reader :client, :abort_download
+
+      # TODO: need to add progress tracker
+      def download(destination, bucket:, **options)
+        if File.exist?(destination)
+          raise ArgumentError 'invalid destination, expected a directory' unless File.directory?(destination)
+        else
+          FileUtils.mkdir_p(destination)
+        end
+
+        download_opts = options.dup
+        @bucket = bucket
+        @ignore_failure = download_opts.delete(:ignore_failure) || false
+        @errors = []
+
+        downloader = FileDownloader.new(client: client, executor: @executor)
+        producer = ObjectProducer.new(destination, build_producer_opts(download_opts))
+        producer.run
+
+        downloads = process_download_queue(producer, downloader, download_opts)
+        build_result(downloads)
+      ensure
+        @executor.shutdown unless @options[:executor]
+      end
+
+      def build_producer_opts(opts)
+        {
+          directory_downloader: self,
+          client: @client,
+          bucket: @bucket,
+          s3_prefix: opts.delete(:s3_prefix),
+          ignore_failure: @ignore_failure,
+          filter_callback: opts.delete(:filter_callback),
+          errors: @errors
+        }
+      end
+
+      def build_result(download_count)
+        downloads = [download_count - @errors.count, 0].max
+
+        if @abort_download
+          msg = "failed to download directory: downloaded #{downloads} files " \
+            "and failed to download #{@errors.count} files."
+          raise DirectoryDownloadError.new(msg, @errors)
+        else
+          result = { completed_downloads: downloads, failed_downloads: @errors.count }
+          result[:errors] = @errors if @errors.any?
+          result
+        end
+      end
+
+      def process_download_queue(producer, downloader, opts)
+        download_attempts = 0
+        completion_queue = Queue.new
+        queue_executor = DefaultExecutor.new
+        while (object = producer.object_queue.shift) != :done
```

Comment:
Suggestion on the Producer interface - I think the object_queue should be an internal detail here.  I'd lean towards having this implement enumerable, so here you would just do:

```
producer.each do |object| 
  break if @abord_download
  # rest of code
end
```

calling `each` on the producer would start it (so no need to call `run`) and would handle yielding objects from the internal object_queue.